File: solutions/java/src/main/java/com/epi/Doors.java
package com.epi;

import java.util.Random;

import static java.lang.Math.floor;
import static java.lang.Math.sqrt;

public class Doors {
  // @include
  static boolean isDoorOpen(int i) {
    double sqrtI = sqrt(i);
    int floorSqrtI = (int)floor(sqrtI);
    return floorSqrtI * floorSqrtI == i;
  }
  // @exclude

  static void checkAnswer(int n) {
    boolean[] doors = new boolean[n + 1]; // false means closed door.
    for (int i = 1; i <= n; ++i) {
      int start = 0;
      while (start + i <= n) {
        start += i;
        doors[start] = !doors[start];
      }
    }

    for (int i = 1; i <= n; ++i) {
      assert isDoorOpen(i) == doors[i];
    }
  }

  public static void main(String[] args) {
    Random gen = new Random();
    int n;
    if (args.length == 1) {
      n = Integer.valueOf(args[0]);
    } else {
      n = gen.nextInt(1000) + 1;
    }
    checkAnswer(n);
  }
}

File: solutions/python/doors.py
import sys
import random
import math


# @include
def is_door_open(i):
    return math.sqrt(i).is_integer()
# @exclude


def check_answer(n):
    doors = [False] * (n + 1)  # False means closed door.
    for i in range(1, n + 1):
        start = 0
        while start + i <= n:
            start += i
            doors[start] = not doors[start]

    for i in range(1, n + 1):
        assert doors[i] == is_door_open(i)


def main():
    n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 1000)
    check_answer(n)


if __name__ == '__main__':
    main()
